### ISSUE TITLE
How to Get Emscripten Bindings to Auto Garbage-Collect References

### ISSUE THREAD

**Issue by zicklag** [emscripten-core/emscripten#8206]
Hello everybody,

I've been using Emscripten with the Haxe [webidl](https://github.com/ncannasse/webidl) library to [automate JS bindings](https://github.com/armory3d/haxebullet/issues/25#issuecomment-462470651) to C++ for [Kha](https://github.com/Kode/Kha/), but I'm needing some help getting the bindings from the webidl library to auto garbage-collect pointers like the Emscripten [webidl binder](https://github.com/emscripten-core/emscripten/blob/incoming/tools/webidl_binder.py).

I'm trying to get Emscripten to work with [Haxebullet](https://github.com/armory3d/haxebullet). We previously used an Ammo.js build, but we wanted to automate the bindings and generate them ourselves. After switching to our own bindings I noticed that the stack was running out of memory because of memory leaks, which were not present in Ammo.js.

I understand from the docs that it may still be necessary to manually delete references because JavaScript will not automatically call destructor methods, but it is not always necessary if there is no destructor to call.

Basically I'm trying to understand, out of what JavaScript wrapping stuff that the Emscripten WebIDL binder does, what is the minimum extra logic and wrapping that is necessary to get JavaScript to auto GC the C++ objects. I don't need to make the API particularly friendly to plain JavaScript because we are using Haxe to wrap the JS in sensibly named objects and functions. The exported JS functions can still be named stuff like `_btCollisionWorld_functionName()` because they will be wrapped in Haxe.

**Comment by kripken** [comment]
If this is the same problem that ammo etc. hit, then it is that a JS object holds an integer, and the integer is to some memory allocated by new/malloc. So we want to free it (and call the destructor if any) when the JS object is GC'd. But we have no way to know when that happens. JS may get weakrefs at some point, which may help, but for now we have to manage this manually.

In ammo and other webidl bindings projects, there is a `destroy()` method that users would need to call on the JS object. That frees the memory and calls the destructor for you.

**Comment by zicklag** [comment]
The weird part is that I actually *didn't* have the same problem with Ammo.js, so I thought that it was doing some kind of auto collecting. If it isn't, then maybe the Ammo's heap is set to growable by default? The only way I knew that it was leaking memory was because the library would out of memory and the app would crash with an exception saying that it could not grow the heap.

**Comment by kripken** [comment]
I think ammo does not have memory growth on by default, but it's in option in the makefile. But I don't remember if that changed at some point. Maybe a build you saw used it.

**Comment by zicklag** [comment]
OK, thanks for the help. :slightly_smiling_face: :+1:
